### PR TITLE
plugins: introduce v2.GoPlugin

### PR DESCRIPTION
**Pull request by sergiusens**
The support added requires the project uses go.mod.
This also adds .get_build_snaps as an abstractmethod to PluginV2.

Signed-off-by: Sergio Schvezov <sergio.schvezov@canonical.com>

- [ ] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `./runtests.sh static`?
- [ ] Have you successfully run `./runtests.sh tests/unit`?

-----

### CODE REVIEW

File: snapcraft/plugins/v2/go.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+class GoPlugin(PluginV2):
+    @classmethod
+    def get_schema(cls) -> Dict[str, Any]:
+        return {
+            "$schema": "http://json-schema.org/draft-04/schema#",
+            "type": "object",
+            "additionalProperties": False,
+            "properties": {
+                "go-channel": {"type": "string", "default": "latest/stable"},
```
Comment (cjp256) [inline snapcraft/plugins/v2/go.py:47]: why about having the user install this with `build-snaps`?

File: tests/spread/plugins/v1/go/snaps/go-hello [hunk 2]
```diff
@@ -0,0 +1 @@
+../../../v2/snaps/go-hello/
```
Comment (cjp256) [inline tests/spread/plugins/v1/go/snaps/go-hello:1]: still linking non-go-mod

File: tests/spread/plugins/v2/build-and-run-hello/task.yaml [hunk 3]
```diff
@@ -20,6 +21,8 @@ restore: |
   # Undo changes to hello
   if [[ "${SNAP}" =~ "python-hello" ]]; then
     git checkout hello
+  elif [[ "${SNAP}" =~ "go-" ]]; then
+      git checkout main.go
```
Comment (cjp256) [inline tests/spread/plugins/v2/build-and-run-hello/task.yaml:25]: nit: indent is different

File: tests/spread/plugins/v2/snaps/go-mod-hello/main.go [hunk 4]
```diff
@@ -21,6 +21,7 @@ package main
 
 import (
 	"fmt"
+	// imported just kick the go.mod tires.
```
Comment (cjp256) [inline tests/spread/plugins/v2/snaps/go-mod-hello/main.go:24]: imported _to_...

File: snapcraft/plugins/v2/go.py (hunk 1 above)
Comment (cjp256) [reply snapcraft/plugins/v2/go.py:47]: ok, I see now that go package support is gone in favor of just the snap channel

Comment (cjp256) [reply snapcraft/plugins/v2/go.py:47]: You may want to add a minimum version check (1.13?)

File: snapcraft/plugins/v2/go.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+                "go-buildtags": {
+                    "type": "array",
```
Comment (cjp256) [inline snapcraft/plugins/v2/go.py:49]: seeing this, I feel like we should gate any plugin properties in core20 on having an associated spread test(s)?

Comment (sergiusens) [reply snapcraft/plugins/v2/go.py:49]: I can add some for this.

Comment (cjp256) [reply snapcraft/plugins/v2/go.py:49]: This could also be generalized "go-extra-install-flags"..

File: snapcraft/plugins/v2/go.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+            "GOBIN": "${SNAPCRAFT_PART_INSTALL}/bin",
+        }
+
+    def get_build_commands(self) -> List[str]:
+        if self.options.go_buildtags:
+            tags = "-tags={}".format(",".join(self.options.go_buildtags))
+        else:
+            tags = ""
+
+        return ["go mod download", f"go install {tags} ${{SNAPCRAFT_GO_LDFLAGS}} ./..."]
```
Comment (cjp256) [inline snapcraft/plugins/v2/go.py:78]: go does support `-p` as the equivalent to make's `-j` for parallel count.  defaults to # of cpus.

File: snapcraft/plugins/v2/go.py (hunk 1 above)
Comment (sergiusens) [reply snapcraft/plugins/v2/go.py:47]: will add the check

File: snapcraft/plugins/v2/go.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+    def get_build_snaps(self) -> Set[str]:
+        return {f"go/{self.options.go_channel}"}
+
+    def get_build_packages(self) -> Set[str]:
+        return {"gcc"}
+
+    def get_build_environment(self) -> Dict[str, str]:
+        return {
+            "SNAPCRAFT_GO_LDFLAGS": "-ldflags -linkmode=external",
```
Comment (cjp256) [inline snapcraft/plugins/v2/go.py:67]: I think this will need to be:
`'-ldflags "-linkmode external"'`

Comment (sergiusens) [reply snapcraft/plugins/v2/go.py:67]: I imported this verbatim from v1.go

File: snapcraft/plugins/v2/go.py (hunk 6 above)
Comment (sergiusens) [reply snapcraft/plugins/v2/go.py:78]: will add

File: snapcraft/plugins/v2/go.py (hunk 1 above)
Comment (sergiusens) [reply snapcraft/plugins/v2/go.py:47]: parsing the channel name you say? I would just let it fail for now

File: snapcraft/plugins/v2/go.py (hunk 7 above)
Comment (cjp256) [reply snapcraft/plugins/v2/go.py:67]: oh i'm a goof, I thought there was a space in there. I think I need a nap :facepalm:

**Pull request merged**